sourcetool/pkg/auth/implementation.go

```diff
@@ -0,0 +1,191 @@
+	for {
+		token, err := di.checkTokenStatus(ctx, deviceCode)
+		if err != nil {
+			if strings.Contains(err.Error(), "authorization_pending") {
+				time.Sleep(interval)
+				continue
+			}
+			if strings.Contains(err.Error(), "slow_down") {
+				interval += 5 * time.Second
+				time.Sleep(interval)
+				continue
+			}
+			return "", err
+		}
+		return token, nil
+	}
+}
+
+// checkTokenStatus checks if the device flow authorization is done and returns
+// the token if ready.
+func (di *defaultImplementation) checkTokenStatus(ctx context.Context, deviceCode string) (string, error) {
+	data := url.Values{}
+	data.Set("client_id", oauthClientID)
+	data.Set("device_code", deviceCode)
+	data.Set("grant_type", "urn:ietf:params:oauth:grant-type:device_code")
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, tokenURL, strings.NewReader(data.Encode()))
+	if err != nil {
+		return "", fmt.Errorf("failed to create token request: %w", err)
+	}
+
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	req.Header.Set("Accept", "application/json")
+
+	httpClient := &http.Client{Timeout: 30 * time.Second}
+	resp, err := httpClient.Do(req)
+	if err != nil {
+		return "", fmt.Errorf("failed to check token status: %w", err)
+	}
+	defer resp.Body.Close() //nolint:errcheck
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return "", fmt.Errorf("failed to read token response: %w", err)
+	}
+
+	var tokenResp TokenResponse
+	if err := json.Unmarshal(body, &tokenResp); err != nil {
+		return "", fmt.Errorf("failed to parse token response: %w", err)
+	}
+
+	if tokenResp.Error != "" {
+		return "", fmt.Errorf("%s", tokenResp.Error)
+	}
+
+	if tokenResp.AccessToken == "" {
+		return "", fmt.Errorf("no access token received")
+	}
+
+	return tokenResp.AccessToken, nil
+}
+
+// openBrowser shells out to open the system's browser to load the
+// device authorization webpage.
+func (di *defaultImplementation) openBrowser(authURL string) error {
+	var cmd string
+	var args []string
+
+	switch runtime.GOOS {
+	case "windows":
+		cmd = "cmd"
+		args = []string{"/c", "start"}
+	case "darwin":
+		cmd = "open"
+	default:
+		cmd = "xdg-open"
+	}
+
+	return exec.Command(cmd, append(args, authURL)...).Start() //nolint:gosec // yes variable input
+}
+
+// requestDeviceCode requests a device code from GitHub
+func (di *defaultImplementation) requestDeviceCode(ctx context.Context) (*DeviceCodeResponse, error) {
+	data := url.Values{}
+	data.Set("client_id", oauthClientID)
+	data.Set("scope", strings.Join(oauthScopes, " "))
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, deviceCodeURL, strings.NewReader(data.Encode()))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create device code request: %w", err)
+	}
+
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	req.Header.Set("Accept", "application/json")
+
+	httpClient := &http.Client{Timeout: 30 * time.Second}
```

Comment:
Same comment here.

Reply:
This is now using the global timeout from the context